test/decision-engine/decision-engine.js

```diff
@@ -0,0 +1,246 @@
+const eachSeries = require('async/eachSeries')
+const CID = require('cids')
+const multihashing = require('multihashing-async')
+const Buffer = require('safe-buffer').Buffer
+
+const Message = require('../../src/types/message')
+const DecisionEngine = require('../../src/decision-engine')
+const createTempRepo = require('../utils/create-temp-repo-nodejs.js')
+
+const mockNetwork = require('../utils/mocks').mockNetwork
+
+function messageToString (m) {
+  return Array.from(m[1].blocks.values())
+    .map((b) => b.data.toString())
+}
+
+function stringifyMessages (messages) {
+  return _.flatten(messages.map(messageToString))
+}
+
+function newEngine (network, callback) {
+  parallel([
+    (cb) => createTempRepo(cb),
+    (cb) => PeerId.create(cb)
+  ], (err, results) => {
+    if (err) {
+      return callback(err)
+    }
+    const blockstore = results[0].blocks
+    const engine = new DecisionEngine(blockstore, network || mockNetwork())
+    engine.start((err) => callback(err, { peer: results[1], engine: engine }))
+  })
+}
+
+describe('Engine', () => {
+  it('consistent accounting', (done) => {
+    parallel([
+      (cb) => newEngine(false, cb),
+      (cb) => newEngine(false, cb)
+    ], (err, res) => {
+      expect(err).to.not.exist()
+
+      const sender = res[0]
+      const receiver = res[1]
+
+      map(_.range(1000), (i, cb) => {
+        const data = Buffer.from(`this is message ${i}`)
+        multihashing(data, 'sha2-256', (err, hash) => {
+          expect(err).to.not.exist()
+
+          const m = new Message(false)
+          const block = new Block(data, new CID(hash))
+          m.addBlock(block)
+          sender.engine.messageSent(receiver.peer, block)
+          receiver.engine.messageReceived(sender.peer, m, cb)
+        })
+      }, (err) => {
+        expect(err).to.not.exist()
+        expect(sender.engine.numBytesSentTo(receiver.peer))
+          .to.be.above(0)
+
+        expect(sender.engine.numBytesSentTo(receiver.peer))
+          .to.eql(receiver.engine.numBytesReceivedFrom(sender.peer))
+
+        expect(receiver.engine.numBytesSentTo(sender.peer))
+          .to.eql(0)
+
+        expect(sender.engine.numBytesReceivedFrom(receiver.peer))
+          .to.eql(0)
+
+        done()
+      })
+    })
+  })
+
+  it('peer is added to peers when message receiver or sent', (done) => {
+    parallel([
+      (cb) => newEngine(false, cb),
+      (cb) => newEngine(false, cb)
+    ], (err, res) => {
+      expect(err).to.not.exist()
+
+      const sanfrancisco = res[0]
+      const seattle = res[1]
+
+      const m = new Message(true)
+      sanfrancisco.engine.messageSent(seattle.peer)
+
+      seattle.engine.messageReceived(sanfrancisco.peer, m, (err) => {
+        expect(err).to.not.exist()
+
+        // TODO why were we expecting for two different peers to have the 
+        // same hexString?
+        // expect(seattle.peer.toHexString())
+        //  .to.eql(sanfrancisco.peer.toHexString())
```

Comment:
@dignifiedquire this assertion started failing when i moved to an async start up but now that I read it, why were we expecting to have different peers with the same Id?